src/chunk.c

```diff
@@ -3054,6 +3063,68 @@ chunk_update_status(FormData_chunk *form)
 							   CurrentMemoryContext) > 0;
 }
 
+/* status update is done in 2 steps.
+ * 1. RowShare lock to read the status.
+ * 2. if status != proposed new status
+ *      update status using RowExclusiveLock
+ * All callers who want to update chunk status should call this function so that locks
+ * are acquired correctly.
+ */
+static bool
+chunk_update_status(FormData_chunk *form)
+{
+	int32 chunk_id = form->id;
+	int32 new_status = form->status;
+	bool success = true, dropped = false;
+	/* lock the chunk tuple for update. Block till we get exclusivetuplelock */
+	ScanTupLock scantuplock = {
+		.waitpolicy = LockWaitBlock,
+		.lockmode = LockTupleExclusive,
+	};
+	ScanIterator iterator = ts_scan_iterator_create(CHUNK, RowShareLock, CurrentMemoryContext);
+	iterator.ctx.index = catalog_get_index(ts_catalog_get(), CHUNK, CHUNK_ID_INDEX);
+	iterator.ctx.tuplock = &scantuplock;
+#if PG12_GE
+	/* see table_tuple_lock for details about flags that are set in TupleExclusive mode */
+	scantuplock.lockflags = TUPLE_LOCK_FLAG_LOCK_UPDATE_IN_PROGRESS;
+	if (!IsolationUsesXactSnapshot())
+	{
+		/* in read committed mode, we follow all updates to this tuple */
+		scantuplock.lockflags |= TUPLE_LOCK_FLAG_FIND_LAST_VERSION;
+	}
+#else
+	scantuplock.lockflags = 1;
```

Comment:
Is this supposed to be `scantuplock.lockflags = TUPLE_LOCK_FLAG_LOCK_UPDATE_IN_PROGRESS`?

Reply:
I see it's being used in `heap_lock_tuple()` for PG version 11. It's really not obvious how it's used in the code. I think we need a second flag fore readability.

Reply:
If you look at heap_lock_tuple, you'll see that this sets follow_updates to true. Not adding another flag for a soon to be deprecated release.

Reply:
I think it makes sense to define `TUPLE_LOCK_FLAG_FIND_LAST_VERSION` as part of the compat lib for PG11 in our own code.